### ISSUE TITLE
visits/orphan-visits day filters do not include the end day as expected

### ISSUE THREAD

**Issue by NReilingh** [shlinkio/shlink-web-client#482]
#### Shlink web client version

* Version: 3.2.0
* How do you use shlink-web-client: self-hosted docker image, app.shlink.io

#### Summary

When observing the visits view for a shortlink or for orphan links, all of the time range filters except for "All visits" exclude the current day's visits, or the selected end date when using a custom range.

#### How to reproduce

1. Create a new short link
2. Visit it
3. Reload the "List short URLs" page, observing a nonzero number of visits for the new shortlink
4. Go to the Visit stats page for that new shortlink

By default, the time range filter on the visit stats page is set to "Last 30 days".

#### Expected behavior

The visit from today should be counted in the default time range of last 30 days; Visits counter should be >0.

#### Current behavior

Visits counter is 0, and the message "There are no visits matching current filter" is displayed

Changing the time range to Last 7 days or today yields the same result. Removing the filter by selecting "All visits" allows today's visits to be counted and displayed. Using the custom range, you can observe that setting the end date to tomorrow will count today's visits, but setting the end date to today will not.

#### Notes

One might also want to filter the time range to a specific month. Under the current behavior, setting start and end dates of the custom range to the beginning and end of the month would result in the last day of the month's visits being excluded.

Is this perhaps a time zone issue? My expectation is that the frontend would be calculating the date range relative to the browser's local time zone, but if this is happening server-side, then my issue could be that my docker container is running in UTC±0. I am currently in UTC-04:00.

**Comment by acelaya** [comment]
I'll take a look.

My guess is that, when calculating the dates, the end date is being set to the beginning of the day instead of the end of the day.

**Comment by acelaya** [comment]
It's actually a combination of what I said, and the fact that you have different timezones in the client and browser.

When using "smart date ranges" (last 7 days, last 30 days, etc), the end date is set to "now", assuming there's not going to be any visit with a date in the future.

However, because of the timezone difference, that is not entirely true.

I'll change the logic to set the end date to "today at 23:59:59", which should solve the issue without any side effects.

**Comment by acelaya** [comment]
I have just merged the workaround. I'm about to release v3.2.1, then you can confirm if it mitigates the problem.

In any case, using the same timezone in the server and client isprobably a good idea.

**Comment by NReilingh** [comment]
@acelaya I've attempted to confirm the fix; unfortunately it did not have the desired effect. I did some more research and I've managed to come to an explanation that I think is plausible.

First, regarding server timezones: it is probably worth noting that all Docker containers, by default, run in UTC, so if time zones are relevant to the server-side, you would want to document this for Docker-based installs. (The easiest way to do this for an Alpine-based image like yours is to define an environment variable like `TZ=America/New_York`, however this will only work if the image has the `tzdata` package installed (`RUN apk add tzdata` in Dockerfile) -- shlink's base image does not currently have this package by default.)

That said, unless I am missing an oblique edge case, shlink appears to store visit datetimes in UTC anyway in all cases, so defining the local time zone for the server should have no effect.

I tested by setting my browser machine's local time zone to UTC, and this immediately caused the expected visits to be filtered properly. This leads me to conclude that the dates are not being compared properly. I took a look at your code change in the attached PR and I don't believe this should have had an effect since you were already using `endDate: new Date()` -- that should have been sufficient.

I'm not sure where this is happening in the code, but I believe what is currently going on is something like:

    Visit UTC date (string) -> Parse as local time string to Date -> Compare with DateRange
    -> Result is out of range when local time zone is in the western hemisphere (UTC<=-1).

What would correct the problem would be

    Visit UTC date (string) -> Parse as UTC time string to Date -> Compare with DateRange
    -> Result is correct in all time zones

This could be as simple as appending `Z` to the ISO date string before parsing it, which will tell the JavaScript date parser that the string is UTC. (The parser defaults to treating ISO dates with ambiguous tz offsets as being relative to the local time zone.) You can demonstrate this in the browser console as follows:

```
> new Date('2021-09-01T14:22:23')
< Wed Sep 01 2021 14:22:23 GMT-0400 (EDT)
> new Date('2021-09-01T14:22:23Z')
< Wed Sep 01 2021 10:22:23 GMT-0400 (EDT)
                   ^ different hour
```

I would guess that you are writing from the eastern hemisphere, which would explain why you are not affected by the issue. :-) Hope this helps provide some useful clues!

**Comment by acelaya** [comment]
Too bad it didn't work. It was an easy change, so it was worth trying, haha.

The problem here is slightly different.

As you guessed, Shlink is using UTC by default, and as you also guessed, I am in an eastern timezone, CET (I live in Spain).

Also, I have my own Shlink instance configured to use my same timezone (it's easy to change on PHP, as it does not use the system timezone, but a configuration file where you can set any value).

So first, I should probably document the fact that you should set your Shlink instance in the same timezone you are going to be accessing to it, although, I need to find a way which matches how other config is provided (right now I'm mounting a file with a volume, which is not ideal).

However, there could be many people allowed to consume the API from different timezones. In this case, you cannot set everyone's timezone.

Because of this, Shlink should probably be smarter, and convert provided dates to the timezone configured in the instance, before running the query.

What is happening right now is that, once dates are persisted in the database, we lose the timezone offset (because of a limitation in some engines and the ORM).

Also, provided dates are formatted without the offset when passed to the database queries.

This makes the possible offset difference between the two to be lost, causing the issue reported here.

So long story short, the logic needs to be fixed in the server.

**Comment by acelaya** [comment]
You can follow the progress here https://github.com/shlinkio/shlink/issues/1175
